## Part 1b: Variables and Data Types

This notebook covers some of the essential Python terminology and concepts. [add more intro stuff] 

#### Variables 
In the previous notebook, you encountered a so-called `variable`. A variable can be seen as a box that holds one or more values. This value can be a number, a string, a list, or any other data type. 

You can give a variable (almost) any name you want. Best practice is to give your variables a name that describes what they are used for. 

The `=` operator is used to assign a value to a variable. Let's look at the example below.

In [21]:
colors = ["red", "green", "blue"]
print(colors)

['red', 'green', 'blue']


In [22]:
pi = 3.14159
print(pi)

3.14159


Python works sequentially, so the order in which you define your variables matters. 

If you use a variable that is not yet defined, you will get an error.

In [23]:
print(soup)
soup = ["broth", "union", "carrots", "celery", "noodles"]

NameError: name 'soup' is not defined

If you give the same name to a variable, only the last value will be stored.

In [25]:
pasta = "bad"
pasta = "good"
print(pasta)

good


#### Data Types
Python makes a distinction between different types of data. The most common data types are:
- `int` for integer numbers (e.g. 1, 2, -3, 486)
- `float` for floating-point numbers (e.g. 0.5, 3.14, -0.0001)
- `str` for strings/text (e.g. "hello", "pasta", "123")
- `bool` for boolean values (True, False)



### Differences between types
Use the function `type()` to check the type of a value or variable.

In [ ]:
type(10)

int

In [ ]:
type(10.10)

float

In [ ]:
type("10")

str

In [ ]:
type(pi)

float

Knowing the data type you're working with is important because they determine what you can do with the data. For example, you can add two integers or floats together, but you can't add an integer and a string. 

In [ ]:
10 + 10.10 

20.1

In [ ]:
10 + "10.10"

TypeError: unsupported operand type(s) for +: 'int' and 'str'

The data type may also determine which functions or methods you can use on the data.

In [ ]:
# You can use the `len()` function to get the length of a string: 
len("beautiful soup")

14

In [ ]:
# But not the length of a number:
len(86)

TypeError: object of type 'int' has no len()

#### Strings

Good to know: `strings` have a lot of built-in methods that you can use to manipulate the data. Let's take a look at some of them.

In [24]:
sentence = "Dua Lipa is an English and Albanian singer, songwriter, and model."

In [ ]:
sentence.lower()

'dua lipa is an english and albanian singer, songwriter, and model.'

In [ ]:
sentence.upper()

'DUA LIPA IS AN ENGLISH AND ALBANIAN SINGER, SONGWRITER, AND MODEL.'

> Exercise 2.1: \
> Predict the output of the following cell. Pay attention to the case of the letters. \
> Did it match you expectations? 

In [ ]:
print(sentence)

> Exercise 2.2: \
> Can you add one line before the print statement to make the output uppercase? \
> (so without changing the print statement)

In [ ]:
# Your code here... 


print(sentence)

#### Booleans 
Let's take a look at Boolean values. These are values that can be either `True` or `False`, and can be useful for making decisions in your code, or for checking if a condition is met. 

In [26]:
5 == 5 

True

In [32]:
1 == 2 

False

Note that `==` is used to signal 'equals'. A single `=` is used to assign a value to a variable.

If you want to check if two values are not equal, you can use `!=`.

In [31]:
6 != 0.5

True

Boolean conditions are often used in `if` statements. These are used to check if a condition is met, and if so, execute a block of code. 

In [37]:
temperature = 25

if temperature > 20:
    print("Eat ice cream")

Eat ice cream


TO DO: 
- excersise with booleans: predict if true 